SharePoint - CONN-3508 - security token (#430)
mulesoft/docs-connectors: modules/ROOT/pages/sharepoint/sharepoint-connector.adoc

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

Support Category: https://www.mulesoft.com/legal/versioning-back-support-policy#anypoint-connectors[Select]

Anypoint Connector for Microsoft SharePoint (Microsoft SharePoint Connector) version 3.1

Release Notes: xref:release-notes::connector/sharepoint-connector-release-notes-mule-4.adoc[Microsoft SharePoint Connector Release Notes - Mule 4]

> Microsoft SharePoint Connector provides support for both SharePoint 2013 and SharePoint Online for use in the cloud and on-premises. This connector enables content and document management, access to intranet and extranet portals and websites, collaboration, and enterprise search. This connector also enables integration with SharePoint through its REST API.

Reviewer: Consider putting the approved short form in the title and the full spelling with short form in parens in the narrative.
It's okay to have abbr in titles as long as the text that follows them spells them out, and doing this helps keep our titles shorter.

Reviewer: "provides support for" > supports?
minimalism

Reviewer: "This connector enables content and document management..." > This connector enables you to manage content and documents, to access..., to collaborate with..., and to search your entire enterprise.
active voice

== Prerequisites

> This document assumes that you are familiar with Microsoft SharePoint, Mule, Anypoint Connectors, Anypoint Studio, Mule concepts, elements in a Mule flow, and Global Elements.

Reviewer: ```suggestion
This document assumes that you are familiar with Microsoft SharePoint, Mule runtime engine (Mule), Anypoint Connectors, Anypoint Studio, Mule concepts, elements in a Mule flow, and Global Elements.
```

You need login credentials to test your connection to your target resource.

For hardware and software requirements and compatibility
information, see the connector Release Notes.

To use this connector with Apache Maven, view the pom.xml dependency information in

> the Dependency Snippets in https://www.mulesoft.com/exchange/com.mulesoft.connectors/mule-sharepoint-connector/[Anypoint Exchange].

Reviewer: Is "Dependency Snippets" the title of a text section or UI element?
Can we give this name a generic noun object, like "text" or "section"?

> == What's New in this Connector

Reviewer: ```suggestion
== What's New in This Connector
```
This and That are always capitalized in titles.

> In Mule 4 SharePoint online and SharePoint on-premises were merged into one connector that includes all functionalities from both previous connectors.

Reviewer: ```suggestion
In Mule 4, SharePoint online and SharePoint on-premises were merged into one connector that includes all functionalities from both previous connectors.
```

== Connect in Design Center

Anypoint Design Center's Flow Designer enables you to create apps visually. For more information, see the xref:design-center::fd-tour.adoc[Flow Designer Tour].

To create an app in Flow Designer:

* Configure the input source (trigger) for your app.

> * Configure the target component for your app.

Reviewer: Are these supposed to be chronological, or does order not matter?

Reviewer: GLOBAL:
Standardize between the more precise of "input source (trigger)" and "input source trigger."
These two texts have DIFFERENT meanings: one implies synonyms and the other a type of trigger.

> For information on Flow Designer field values, see the xref:connectors::sharepoint/sharepoint-connector-reference.adoc[Microsoft SharePoint Connector Reference].

Reviewer: GLOBAL:
information on > information about

> == Configure the Input Source Trigger

Reviewer: See previous comment about this wording.
Please fix in template, if this appears there.

. In Design Center, click *Create*.
. Click *Create new application* to open Flow Designer.

> . Specify a value for *Project name*, and click *Create*.

Reviewer: ```suggestion
. Specify a value for *Project name* and click *Create*.
```

. Click *Go straight to canvas* to exit from *Let's get started*. 
. Click on the trigger card.

> . If you are using your connector as an input source trigger, search for the connector; otherwise, you can use HTTP or Scheduler.

Reviewer: ditto re. term

Reviewer: "you can use" > "use"
minimalism

+
Microsoft SharePoint Connector provides trigger operations for:

* Created Objects
* Modified Objects

> * Deleted Objects

Reviewer: There is no reason to capitalize "Objects" in these list items.

+
image::sharepoint/sharepoint-global-config.png[]
+

> . Provide these required fields (differs by operation):

Reviewer: ```suggestion
. Depending on the operation you are performing, provide these required fields:
```
To ease translation, avoid parentheticals when possible.

.. Object Type: Folder, File, List, or List Item
.. Connection: Okta, Online, or security token authentication
.. Security Token
.. Site URL

> . Select the plus sign to add a component

Reviewer: ```suggestion
. Select the plus sign (+) to add a component.
```
internationalization

> . Select the connector as a component

Reviewer: period

> . Configure the component

Reviewer: period

* Okta API Token: Okta uses a bearer token for API authentication with a sliding scale expiration, for example: `+http://developer.okta.com/docs/api/getting_started/getting_a_token.html+`
* SharePoint Online Embedded Link: The link of your Microsoft Office 365 SharePoint Online application from Okta, for example: `+https://your-domain.okta.com/home/office365/0oa14mz7f4HAWypdZ1t7/31737+`

== Security Token Authentication

> You can use a SAML security token to log into SharePoint.

Reviewer: GLOBAL:

to login to > to log in to
(login is an adjective or a noun, but not a verb)

Reviewer: Substitute colon for period.

image::sharepoint/sharepoint-security-token-config.png[Security Token Configuration]

> * Security token: Enter the coresponding security token.

Reviewer: ```suggestion
* Security token: Enter the corresponding security token.
```

> * Site URL: Enter the coresponding site URL.

Reviewer: ```suggestion
* Site URL: Enter the corresponding site URL.
```

Reviewer: GLOBAL:
corresponding

> A security token can be obtained by making a post request to `https://login.microsoftonline.com/extSTS.srf` with the following XML body:

Reviewer: ```suggestion
You can obtain a security token by making a post request to `https://login.microsoftonline.com/extSTS.srf` with the following XML body:
```
active voice

    <a:ReplyTo>
      <a:Address>http://www.w3.org/2005/08/addressing/anonymous</a:Address>
    </a:ReplyTo>
    <a:To s:mustUnderstand="1">https://login.microsoftonline.com/extSTS.srf</a:To>
    <o:Security s:mustUnderstand="1"
       xmlns:o="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-secext-1.0.xsd">
      <o:UsernameToken>
        <o:Username>[username]</o:Username>
        <o:Password>[password]</o:Password>
      </o:UsernameToken>
    </o:Security>
  </s:Header>
  <s:Body>
    <t:RequestSecurityToken xmlns:t="http://schemas.xmlsoap.org/ws/2005/02/trust">
      <wsp:AppliesTo xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy">
        <a:EndpointReference>
          <a:Address>[endpoint]</a:Address>
        </a:EndpointReference>
      </wsp:AppliesTo>
      <t:KeyType>http://schemas.xmlsoap.org/ws/2005/05/identity/NoProofKey</t:KeyType>
      <t:RequestType>http://schemas.xmlsoap.org/ws/2005/02/trust/Issue</t:RequestType>
      <t:TokenType>urn:oasis:names:tc:SAML:1.0:assertion</t:TokenType>
    </t:RequestSecurityToken>
  </s:Body>
</s:Envelope>
----

> The response from the above request contains the security token between the <wsse:BinarySecurityToken> tags.

Reviewer: ```suggestion
The response from this request contains the security token between the <wsse:BinarySecurityToken> tags.
```
GLOBALLY replace references to "above" and "below," per the style reference.

You can use this connector in Anypoint Studio by first downloading it from Exchange
and configuring it as needed.

== Add the Connector to Your Studio Project

> Anypoint Studio provides two ways to add the connector to your Studio project: from the Exchange button in the Studio taskbar or from the Mule Palette.

Reviewer: ```suggestion
Anypoint Studio provides two ways to add the connector to your Studio project: from the Exchange button in the Studio taskbar or from the Mule Palette view.
```
PLEASE UPDATE THE TEMPLATE.

=== Add the Connector Using Exchange

. In Studio, create a Mule project.
. Click the Exchange icon *(X)* in the upper-left of the Studio task bar.
. In Exchange, click *Login* and supply your Anypoint Platform username and password.
. In Exchange, search for "share".
. Select the connector and click *Add to project*.
. Follow the prompts to install the connector.

=== Add the Connector in Studio

. In Studio, create a Mule project.

> . In the Mule Palette view, click *(X) Search in Exchange*.

Reviewer: YAY view!

. In *Add Modules to Project*, type "share" in the search field.
. Click the connector's name in *Available modules*.
. Click *Add*.
. Click *Finish*.

=== Configure Your First Flow

To configure:

. In Anypoint Studio, click *File* > *New* > *Mule Project*.

> . Specify a *Project Name* and click *Finish*.

Reviewer: ```suggestion
. Specify a value for *Project Name* and click *Finish*.
```

. Click the *Global Elements* tab.
. Click *Create*.
. Search for "sharepoint".
. Click *Microsoft SharePoint* and click *OK*.
. Choose the Global Type to configure:
+
image::sharepoint/sharepoint-global-types.png[Global Types]
+
. Click *Test Connection* to ensure the connection works correctly:
+
image::sharepoint/sharepoint-global-element-props.png[Global Element Properties]

=== Create an Anypoint Studio Flow

To create an Anypoint Studio flow:

. In Anypoint Studio, click *File* > *New* > *Mule Project*.

> . Specify a *Project Name* and click *Finish*.

Reviewer: ```suggestion
. Specify a value for *Project Name* and click *Finish*.
```

. Search for `http` and drag an HTTP connector to the canvas.
. Search for `sharepoint` and drag a Microsoft SharePoint connector next to the HTTP connector.
+
image::sharepoint/sharepoint-studio-flow.png[Studio Visual Editor Flow]
+

> . Double-click the HTTP connector and set *Host* to `0.0.0.0`, *Port* to `8081`, and *Path* to `fileAdd`. Click *OK*.

Reviewer: Two steps:
```suggestion
. Double-click the HTTP connector and set *Host* to `0.0.0.0`, *Port* to `8081`, and *Path* to `fileAdd`. 
. Click *OK*.
```

* Anypoint Studio with at least the Mule 4.0 Runtime.
* Microsoft SharePoint Connector v2.0.0 or higher.

=== Test the Flow

> . Import the demo project into your workspace using Anypoint Exchange or using the *File* > *Import command.

Reviewer: ```suggestion
. Import the demo project to your workspace using Anypoint Exchange or using the *File* > *Import command.
```
NOTE: there is something that needs fixing in "the *File* > *Import command". 
Maybe just a missing *
But...clicking UI elements is not using a command....

. Specify your credentials in the `/src/main/app/mule-app.properties` file:
. Run the project in Studio.
. Type `+localhost:8081/demo+` in your browser to access the selection menu of the demo.
. Optionally you can configure *Connection Timeout* and *Read Timeout*.
*Connection Timeout* is the timeout in making the initial connection with the server.

> *Read Timeout* is the timeout on waiting to read data from the server.

Reviewer: Instead of defining "timeout" with "timeout," consider...:

*Connection Timeout* is the maximum time allowed to make the initial server connection.
*Read Timeout* is the maximum time allowed to read data from the server.

> You can use the selection menu from `+http://localhost:8081+` to test the flows or you can POST JSONs using a tool like curl, or any other tool (Chrome/Mozilla Firefox extensions) that lets you post an HTTP body when calling the URL.

Reviewer: ```suggestion
You can either use the selection menu from `+http://localhost:8081+` to test the flows, or you can POST JSONs using a tool like curl, or any other tool (Chrome/Mozilla Firefox extensions) that lets you post an HTTP body when calling the URL.
```

+
* What authentication schemes are supported by this connector?
+

> For SharePoint Online, authentication using standard SharePoint online user credentials is supported, secure authentication,  as well as Okta authentication.

Reviewer: ```suggestion
SharePoint Online supports authentication using standard SharePoint Online user credentials, secure authentication, and Okta authentication.
```
